File: algorithms/0047.permutations_ii/permutations_ii.cpp
class Solution {
public:
    void backtracking(const vector<int> &nums, vector<bool> &used) {
        if (path_.size() == nums.size()) {
            permute_.push_back(path_);
            return;
        }
        for (int i = 0; i < nums.size(); ++i) {
            if (i > 0 && nums[i-1] == nums[i] && used[i-1] == false) {
                continue;
            }
            if (used[i] == false) {
                used[i] = true;
                path_.push_back(nums[i]);
                backtracking(nums, used);
                used[i] = false;
                path_.pop_back();
            }
        }
    }
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<bool> used(nums.size(), false);
        backtracking(nums, used);
        return permute_;
    }
private:
    vector<vector<int>> permute_;
    vector<int> path_;
};

File: algorithms/0047.permutations_ii/permutations_ii.rs
impl Solution {
    pub fn permute_unique(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut nums = nums.to_owned();
        let mut path = Vec::new();
        let mut visted = vec![false; nums.len()];
        let mut permutations = Vec::new();
        nums.sort();
        Self::backtracking(&nums, &mut path, &mut permutations, &mut visted);
        permutations
    }
    fn backtracking(nums: &Vec<i32>, path: &mut Vec<i32>, permutations: &mut Vec<Vec<i32>>, visted: &mut Vec<bool>) {
        if path.len() == nums.len() {
            permutations.push((*path).clone());
            return;
        }
        for i in 0..nums.len() {
            if i != 0 && nums[i] == nums[i - 1] && !visted[i - 1] {
                continue;
            }
            if !visted[i] {
                path.push(nums[i]);
                visted[i] = true;
                Self::backtracking(nums, path, permutations, visted);
                path.pop();
                visted[i] = false;
            }
        }
    }
}